samples/cube.lobster

```diff
@@ -0,0 +1,111 @@
+// This example renders a cube with a different colorèd lobster on each face.
+// Moving the mouse or WASD rotates the viewport around the cube like turning a
+// globe in front of the camera.
+//
+// The example illustrates creating a simple mesh and rendering a texture on
+// the mesh in various orientations.
+// The example showcases the render_to_texture utility, which allows us to
+// render a 2D drawing on the surfaces of a 3D object, using a frame buffer.
+//
+// Kris Kowal <kris@cixar.com>
+
+import vec
+import color
+import util3d
+
+fatal(gl_window("Lobster Cube", 515, 515))
+
+// Generate a mesh for a unit square.
+let square_corners = [ xy_0, xy_x, xy_1, xy_y, xy_0 ]
+let square_mesh = gl_new_poly(square_corners)
+
+// How to break glass when the texture coordinates do not match the mesh
+// vertices:
+// let square_positions = map(square_corners) v: xyz(v, 0)
+// let square_texcoords1 = square_corners
+// let square_indices = [0, 1, 2, 2, 3, 0]
+// let square_mesh = gl_new_mesh(
+//     "PT",
+//     square_positions,
+//     [], // square_colors,
+//     [], // square_normals,
+//     square_texcoords1,
+//     [], // square_textcoords2,
+//     square_indices
+// )
+
+let lobster_tex = gl_load_texture("data/textures/lobster.jpg")
+assert lobster_tex
+
+let detail = 256
+let face_tex = map([
+    color_purple,
+    color_olive,
+    color_teal,
+    color_dark_red,
+    color_dark_blue,
+    color_dark_green,
+]) color:
+    render_to_texture(nil, xy_1i * detail, false, nil, 0):
+        gl_set_shader("textured")
+        gl_set_primitive_texture(0, lobster_tex)
+        // The texture colors are multiplied by the color in context.
+        // Since our texture is white on black, we can change the white to a
+        // unique color for each face of the world.
+        gl_color(color)
+        gl_rect(xy_1 * detail, 0)
+
+def transform_to_face(i, f):
+    gl_translate(xyz_0):
+        if i == 0: // -x
+            gl_translate(xyz_x / -2)
+            gl_rotate_z(-xy_y)
+            gl_rotate_x(-xy_y)
+            gl_translate((xyz_x + xyz_y) / -2)
+        else: if i == 1: // -y
```

Comment:
wonder if theres a more elegant way to do this. in particular, the `lobstercraft.lobster` achieves this by simply generating 6 faces with different params.

Reply:
I’d like to use an array of transform matrices. That would eliminate the function entirely.

I saw the trick in lobstercraft too. Now that I better understand how it works, I might have more luck. It would be interesting to see how it treats the orientation of the textures. One of the motivations of this approach was versatility. From this jump-off, I can render a selection box on the surface on top of the cube, for example. Or, dynamically render the surface texture on the frame buffer. One way I can run with this is to do a CA demo on the cube surface.

Reply:
To create such matrices, you'd still need functions similar to the ones you're calling, so it wouldn't do much for code complexity. Generally though, rendering something like this with transforms doesn't seem like the way to go.

Reply:
I’ve adapted the solution from LobsterCraft™, and basically wrote my dissertation on reverse engineering it. Hope that helps.